frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -509,94 +539,120 @@ static void checkForParenlessMethodFieldRedefinition(Context* context,
-      auto parentFn = parentAst->toFunction();
-      parentFnUntyped = UntypedFnSignature::get(context, parentFn);
-      parentFnTyped = typedSignatureInitial(context, parentFnUntyped);
+  if (fn == nullptr) return nullptr;
+
+  // Construct the parent function's signature.
+  // If the input is a nested function, then a frame for the parent signature
+  // may be present in the 'ResolutionContext'. If no frame is present, then
+  // attempt to call 'typedSignatureInitial' on the parent. This will give
+  // up if any parent is generic or if any parent contains outer variables.
+  const TypedFnSignature* parentSignature = nullptr;
+  if (ID parentFnId = fn->id().parentFunctionId(context)) {
+    if (auto frame = rc->findFrameWithId(parentFnId)) {
+      if (auto sig = frame->signature()) {
+        CHPL_ASSERT(sig->id() && sig->id().contains(fn->id()));
+        parentSignature = sig;
+      }
+    } else {
+      auto parentShape = UntypedFnSignature::get(context, parentFnId);
+      parentSignature = typedSignatureInitial(rc, parentShape);
     }
 
-    ResolutionResultByPostorderID r;
-    auto visitor = Resolver::createForInitialSignature(context, fn, r);
-    // visit the formals
-    for (auto formal : fn->formals()) {
-      formal->traverse(visitor);
+    // The parent signature must exist.
+    if (!parentSignature) return nullptr;
+
+    // All parents need to be concrete for the signature to be evaluated.
+    for (auto up = parentSignature; up; up = up->parentFn()) {
+      if (up->needsInstantiation()) return nullptr;
     }
-    // do not visit the return type or function body
+  }
 
-    // now, construct a TypedFnSignature from the result
-    std::vector<types::QualifiedType> formalTypes = visitor.getFormalTypes(fn);
-    bool needsInstantiation = anyFormalNeedsInstantiation(context, formalTypes,
-                                                          untypedSig,
-                                                          nullptr);
+  ResolutionResultByPostorderID r;
+  auto visitor = Resolver::createForInitialSignature(rc, fn, r);
 
-    // visit the where clause, unless it needs to be instantiated, in
-    // which case we will visit the where clause when that happens
-    TypedFnSignature::WhereClauseResult whereResult =
-      TypedFnSignature::WHERE_NONE;
-    if (auto whereClause = fn->whereClause()) {
-      if (needsInstantiation) {
-        whereResult = TypedFnSignature::WHERE_TBD;
-      } else {
-        whereClause->traverse(visitor);
-        whereResult = whereClauseResult(context, fn, r, needsInstantiation);
-      }
-    }
+  // visit the formals, but not the return type or body
+  for (auto formal : fn->formals()) formal->traverse(visitor);
 
-    checkForParenlessMethodFieldRedefinition(context, fn, visitor);
+  // Give up if we could not type outer variables present in the signature.
+  if (parentSignature && rc->isEmpty() && !visitor.outerVariables.isEmpty()) {
+    return nullptr;
```

Comment:
Wouldn't an error need to be emitted in this case? Who is responsible for emitting the error?

Reply:
Seems more like an API issue (stack frames not present) but we could emit some sort of error indicating as such.

Reply:
OK, in that event, just put a comment saying you expect this can only happen when it is called incorrectly